adoc/chapters/device_compiler.adoc

```diff
@@ -216,132 +216,31 @@ The restriction waiver in <<discarded-statement>> or
-a device compiler command-line option to ensure the types are the same.
+The following types must be available in device code:
 
-The standard {cpp} fixed width types, e.g. [code]#int8_t#, [code]#int16_t#,
-[code]#int32_t#,[code]#int64_t#, should have the same size as defined by the
-{cpp} standard for host and device.
+- The fundamental data types defined by the {cpp} core language (i.e.,
+  [code]#bool#, [code]#char#, [code]#signed char#, [code]#unsigned char#,
+  [code]#short int#, [code]#unsigned short int#, [code]#int#, [code]#unsigned
+  int#, [code]#long int#, [code]#unsigned long int#, [code]#long long int#,
+  [code]#unsigned long long int#, [code]#float# and [code]#double#);
 
+- The fixed width types defined by the {cpp} core language (e.g.,
+  [code]#int8_t#, [code]#int16_t#, [code]#int32_t#, [code]#int64_t#); and
 
-[[table.types.fundamental]]
-.Fundamental data types supported by SYCL
-[width="100%",options="header",separator="@",cols="65%,35%"]
-|====
-@ Fundamental data type @ Description
-a@
-[source]
-----
-bool
-----
-   a@ A conditional data type which can be either true or false. The value
-      true expands to the integer constant 1 and the value false expands to the
-      integer constant 0.
-
-a@
-[source]
-----
-char
-----
-   a@ A signed or unsigned 8-bit integer, as defined by the {cpp} core language
-
-a@
-[source]
-----
-signed char
-----
-   a@ A signed 8-bit integer, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned char
-----
-   a@ An unsigned 8-bit integer, as defined by the {cpp} core language
-
-a@
-[source]
-----
-short int
-----
-   a@ A signed integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned short int
-----
-   a@ An unsigned integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-int
-----
-   a@ A signed integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned int
-----
-   a@ An unsigned integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-long int
-----
-   a@ A signed integer of at least 32-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned long int
-----
-   a@ An unsigned integer of at least 32-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-long long int
-----
-   a@ An integer of at least 64-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned long long int
-----
-   a@ An unsigned integer of at least 64-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-float
-----
-   a@ A 32-bit floating-point. The float data type must conform to the IEEE 754
-      single precision storage format.
-
-a@
-[source]
-----
-double
-----
-   a@ A 64-bit floating-point. The double data type must conform to the IEEE 754
-      double precision storage format.  This type is only supported on devices
-      that have [code]#aspect::fp64#.
```

Comment:
Removing this loses the requirement that `float` and `double` correspond to the IEEE 754 specification. C++ doesn't restrict these types to IEEE 754 (there are C++ implementations where these types are not IEEE 754 by default). If it is intentional that SYCL strengthens the requirements relative to the C++ standard, then this still needs to be stated somewhere. I wouldn't be surprised if these requirements were not intended to deviate from the C++ standard; most C++ programmers tend to think of these types as IEEE 754 types. I'm not sufficiently familiar with their use in SYCL to have an opinion on whether this requirement is warranted.

Reply:
There was some discussion of this in internal issue 322.

This tie to IEEE 754 seems to have originated in OpenCL.  Almost every other behavior that was inherited from OpenCL has been relaxed, deprecated, or removed entirely to enable SYCL to be implemented on top of other backends.  It strikes me as odd that we've kept this restriction, especially given that C++ doesn't require it.

@gmlueck, you were involved in internal issue 322.  What do you think we should do here?  Can we relax this requirement in SYCL-Next?  Or is it something we should have removed when transitioning between SYCL 1.2.1 and SYCL 2020?

Reply:
Yes, it seems weird to me also.  I think it would make sense to remove the requirement that `float` and `double` have IEEE 754 storage format.